File: string_factor.cpp
/*
    brief: 字符串相乘
    time: 20190312
    tips:
*/

#include <string>
#include <vector>
#include <algorithm>
#include <stdio.h>
#include <iostream>

std::string GetFac(const std::string &s1, const std::string &s2)
{
    // check param
    if(s1.empty() || "0" == s1 || s2.empty() || "0" == s2)
    {
        return "0";        
    }

    // reverse
    std::string sF1(s1.rbegin(), s1.rend());
    std::string sF2(s2.rbegin(), s2.rend());

    // 符号
    bool isNeg = false;
    if('-' == s1[0] && '-' != s2[0] || '-' == s2[0] && '-' != s1[0])
    {
        isNeg = true;
    }

    // 相乘
    std::vector<int> viRes(s1.size() + s2.size(), 0);
    for(int i = 0; i < sF1.size(); ++i)
    {
        if(sF1[i] < '0' || sF1[i] > '9')
        {
            break;
        }
        int f1 = sF1[i] - '0';
        for(int j = 0; j < sF2.size(); ++j)
        {
            if(sF2[j] < '0' || sF2[j] > '9')
            {
                break;
            }
            viRes[i + j] += (sF2[j] - '0') * f1;
        }
    }

    // 处理最后结果
    std::string sRes = "";
    int carry = 0;
    for(int i = 0 ; i < viRes.size(); ++i)
    {
        viRes[i] += carry;
        if(viRes[i] >= 10)
        {
            carry = viRes[i] / 10;
            viRes[i] %= 10;
        }
        else
        {
            carry = 0;
        }
        sRes.push_back(viRes[i] + '0');
    }

    // 最后整理
    size_t pos = sRes.find_last_not_of("0");
    if(std::string::npos == pos)
    {
        return "0";
    }
    sRes.erase(pos + 1);
    if(isNeg)
    {
        sRes.push_back('-');
    }
    std::reverse(sRes.begin(), sRes.end());
    return sRes;
}

int main()
{
    std::string s1, s2;
    while(std::cin >> s1 >> s2)
    {
        printf("%s * %s = %s\n", s1.c_str(), s2.c_str(), GetFac(s1, s2).c_str());        
    }
}

File: string_factor.go
/*
    brief: 字符串相乘 golang版本
    time: 20190312
    tips:
*/

package main

import (
    "fmt"
)

func ReverseString(s string) string {
    if("" == s){
        return ""
    }
    sb := []byte(s)
    for i, j := 0, len(sb) - 1; i < j; i, j = i + 1, j - 1 {
        sb[i], sb[j] = sb[j], sb[i]
    }
    return string(sb)
}

func GetFac(s1, s2 string) string{
    // check param
    if "" == s1 || "0" == s1 || "" == s2 || "0" == s2{
        return "0"      
    }

    // 符号
    isNeg := false
    if '-' == s1[0] && '-' != s2[0] || '-' == s2[0] && '-' != s1[0]{
        isNeg = true
    }

    // reverse
    s1 = ReverseString(s1)
    s2 = ReverseString(s2)

    // 相乘
    resArray := make([]int, len(s1)+len(s2))
    for i := 0; i < len(s1); i ++ {
        if '+' == s1[i] || '-' == s1[i] {
            break
        }
        f1 := int(s1[i] - '0')
        for j := 0; j < len(s2); j ++ {
            if '+' == s2[j] || '-' == s2[j] {
                break;
            }
            resArray[i + j] += int(s2[j] - '0') * f1
        }
    }

    // 处理最后结果
    resString := make([]byte, len(resArray))
    carry := 0
    for index, val := range resArray {
        val += carry
        if val >= 10 {
            carry = val / 10
            val %= 10
        }else{
            carry = 0
        }
        resString[index] = byte(val + '0')
    }

    // 最后整理
    i := len(resString) - 1
    for ; i >= 0; i-- {
        if resString[i] != '0' {
            break
        }
    }
    if isNeg {
        i++
        resString[i] = '-'
    }
    
    return ReverseString(string(resString[0 : i+1]));
}

func main(){
    var s1, s2 string
    for{
        fmt.Scanf("%s", &s1)
        fmt.Scanf("%s", &s2)
    
        resString := GetFac(s1, s2)
        fmt.Printf("factor:%s\n", resString)
    }
    
}
